core/src/test/scala/kanaloa/reactive/dispatcher/queue/QueueProcessorSpec.scala

```diff
@@ -0,0 +1,233 @@
+    "scale workers down" in withQueueProcessor() { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
+
+      qp ! ScaleTo(3) //kill 2 Workers
+
+      //can't think of a better way to do this right now.
+      val retiredProbes = workerFactory.probeMap.values.filter { probe ⇒
+        Try { probe.expectMsg(Worker.Retire) }.toOption.isDefined
+      }
+
+      retiredProbes should have size 2
+
+      //kill the 'Workers' who got the Retire message, so that they signal the QP to remove them
+      retiredProbes.foreach(_.ref ! PoisonPill)
+
+      eventually {
+        qp.underlyingActor.workerPool should have size 3
+      }
+    }
+
+    "honor minimum pool size during AutoScale" in withQueueProcessor() { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
+      qp ! ScaleTo(1) //minimum is 3
+
+      //can't think of a better way to do this right now.
+      val retiredProbes = workerFactory.probeMap.values.filter { probe ⇒
+        Try { probe.expectMsg(Worker.Retire) }.toOption.isDefined
+      }
+
+      //kill the 'Workers' who got the Retire message, so that they signal the QP to remove them
+      retiredProbes.foreach(_.ref ! PoisonPill)
+
+      eventually {
+        qp.underlyingActor.workerPool should have size 3
+      }
+    }
+
+    "honor maximum pool size during AutoScale" in
+      withQueueProcessor(ProcessingWorkerPoolSettings(maxPoolSize = 7)) { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
+        qp ! ScaleTo(10) //maximum is 7
+
+        eventually {
+          qp.underlyingActor.workerPool should have size 7
+          testBackend.timesInvoked shouldBe 7
+        }
+      }
+
+    "shutdown Queue and Workers" in withQueueProcessor() { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
+
+      qp ! Shutdown(Some(self), 30.seconds)
+      queueProbe.expectMsg(Queue.Retire(30.seconds))
+
+      qp ! QueryStatus()
+      expectMsg(ShuttingDown)
+
+      //when the Queue is told to shutDown, it will send
+      //term signals to Workers.  Workers will then eventually terminate
+      //These PoisonPills simulate that
+
+      workerFactory.probeMap.values.foreach { probe ⇒
+        probe.ref ! PoisonPill
+      }
+
+      expectMsg(ShutdownSuccessfully)
+      expectTerminated(qp)
+    }
+
+    "shutdown if Queue terminates" in withQueueProcessor() { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
+
+      queueProbe.ref ! PoisonPill
+
+      workerFactory.probeMap.values.foreach { probe ⇒
+        probe.expectMsg(Worker.Retire)
+      }
+
+      qp ! QueryStatus()
+      expectMsg(ShuttingDown)
+
+      //simulate the Workers all finishing up
+      workerFactory.probeMap.values.foreach { probe ⇒
+        probe.ref ! PoisonPill
+      }
+
+      expectTerminated(qp)
+
+    }
+
+    //TODO: not sure why this is broken.
+    "force shutdown if timeout" ignore withQueueProcessor() { (qp, queueProbe, metricsCollector, testBackend, workerFactory) ⇒
```

Comment:
No clue why this is failing.  It just hangs, and I still need to figure out what is going on.